classes/boardmanager.php

```diff
@@ -0,0 +1,1041 @@
+        $id = $this->create_board_from_template($this->board->id, ['template' => 1]);
+        $this->formatter->put('common', ['template' => $id]);
+        return $id;
+    }
+
+    /**
+     * Creates a board for the whole course.
+     *
+     * @return int Id of the new board
+     */
+    public function create_board(): int {
+        return $this->create_board_from_template($this->get_template_board_id(), ['userid' => 0, 'groupid' => 0]);
+    }
+
+    /**
+     * Creates a new board from a template. If no template is given or found, the default template is used.
+     * Assigned users, discussions and history are not copied.
+     *
+     * @param int $templateid Board id of the template.
+     * @param array $data Data to override in the board record
+     * @return int Id of the new board
+     */
+    public function create_board_from_template(int $templateid = 0, array $data = []): int {
+        global $DB;
+        if (empty($templateid)) {
+            $templateid = $this->get_template_board_id();
+        }
+        // Template can still not exist (if kanban instance has none). Use default template.
+        if (empty($templateid)) {
+            $boarddata = [
+                'sequence' => '',
+                'userid' => 0,
+                'groupid' => 0,
+                'template' => 0,
+                'timecreated' => time(),
+                'timemodified' => time(),
+                'kanban_instance' => $this->kanban->id
+            ];
+            // Replace / append data.
+            $boarddata = array_merge($boarddata, $data);
+            $boardid = $DB->insert_record('kanban_board', $boarddata);
+            $columnnames = [
+                get_string('todo', 'kanban') => '{}',
+                get_string('doing', 'kanban') => '{}',
+                get_string('done', 'kanban') => '{"autoclose": true}',
+            ];
+            $columnids = [];
+            foreach ($columnnames as $columnname => $options) {
+                $columnids[] = $DB->insert_record('kanban_column', [
+                    'title' => $columnname,
+                    'sequence' => '',
+                    'kanban_board' => $boardid,
+                    'options' => $options,
+                    'timecreated' => time(),
+                    'timemodified' => time(),
+                ]);
+            }
+            $DB->update_record('kanban_board', ['id' => $boardid, 'sequence' => join(',', $columnids)]);
+            helper::update_cached_board($boardid);
+            return $boardid;
+        } else {
+            $fs = get_file_storage();
+            $template = get_cached_board($templateid);
+
+            // If it is a site wide template, we need system context to copy files.
+            if ($template->kanban_instance == 0) {
+                $context = context_system::instance(0);
+            } else {
+                $context = context_module::instance($this->cmid, 'kanban');
```

Comment:
The retrieving of the context should be moved down to the place where it is actually needed